### PR TITLE
Table-based formatting of entries in `read ledger.py`

### PR DESCRIPTION
**Pull request by jumaffre**
~`read_ledger.py` didn't cope with `-1` keys (use in the new configuration table). This PR fixes it and also runs the utility in the CI so that we notice if it breaks again.~

We now format the keys and values of each table depending on the name of this one.

### CODE REVIEW

File: python/ccf/read_ledger.py [hunk 1]
```diff
@@ -15,13 +15,17 @@ def indent(n):
...
-    s = bs.decode()
-    if s.isprintable():
-        return s
-    if len(bs) > 0 and len(bs) <= 8:
-        n = int.from_bytes(bs, byteorder="little")
-        return f"<u{8 * len(bs)}: {n}>"
-    return bs
+    try:
+        s = bs.decode()
+        if s.isprintable():
```
Comment (achamayou) [inline python/ccf/read_ledger.py:20]: I'm not convinced this is a good way to pretty print the contents of arbitrary buffers, it will sometimes display short, unprintable sequences as unsigned integers (signed, with a u prefix!?), and printable integers will be shown as strings. I would prefer if we defaulted to something more regular, even if that means having to rely on schema hints to make things nicer for governance.

Comment (wintersteiger) [reply python/ccf/read_ledger.py:20]: Yeah, we would need some sort of type hints to make this nicer. We could also just print it in hex.

Comment (achamayou) [reply python/ccf/read_ledger.py:20]: We could embed hints about governance tables, because we have to keep that schema extremely stable anyway, and allow users to supply their own hints as well. Defaulting to hex is reasonable, I agree.

**Comment by achamayou** [comment]
As discussed, let's go with:

1. Print to keys and values as  hex
2. For public:ccf.gov.*, print keys as strings, values as strings
3. For public:ccf.gov.[tables with unsigned int keys], print keys as unsigned ints

File: tests/suite/test_suite.py [hunk 2]
```diff
@@ -107,13 +107,17 @@
...
-    governance_history.test_tables_doc,
     # curve migration:
     reconfiguration.test_change_curve,
     recovery.test,
     # jwt
     jwt_test.test_refresh_jwt_issuer,
+    #
+    #
+    #
+    # Below tests should always stay last
```
Comment (ashamis) [inline tests/suite/test_suite.py:118]: It might be useful to extend the comment explaining why this is the case

Comment (achamayou) [reply tests/suite/test_suite.py:118]: For test_tables_doc, to goal is to collect all the tables the tests have used, parse the doc, and complain if they aren't mentioned. So naturally, to avoid missing tables, you want it last.
For the ledger is readable, it's the same idea, you want to test as much of it as possible.

Neither test will break if it's not last, but they will be less useful.

**Comment by jumaffre** [comment]
> As discussed, let's go with:
> 
> 1. Print to keys and values as  hex
> 2. For public:ccf.gov.*, print keys as strings, values as strings
> 3. For public:ccf.gov.[tables with unsigned int keys], print keys as unsigned ints
> 
> The top-level call can take a list of (regexp, rendering method) that users can extend/override.

This should now be done and ready for review.

File: python/ccf/read_ledger.py [hunk 3]
```diff
@@ -14,18 +14,49 @@ def indent(n):
...
+        "key": fmt_unsigned_int_little,
+        "value": fmt_json,
+    },
+    "^public:ccf\.gov\.(service|network|constitution).*$": {
+        "key": fmt_unsigned_int_little,
+        "value": fmt_json,
+    },
+    "^public:ccf\.gov\..*$": {"key": fmt_str, "value": fmt_json},
+    ".*": {"key": fmt_hex, "value": fmt_hex},
+}
```
Comment (eddyashton) [inline python/ccf/read_ledger.py:48]: I think this should be a list of pairs, rather than a dict. We don't need random lookup, we just iterate, and that makes it clearer that the order is intentional and meaningful.

File: python/ccf/read_ledger.py [hunk 4]
```diff
@@ -14,18 +14,51 @@ def indent(n):
...
 
-def print_key(indent_s, k, is_removed=False):
-    k = stringify_bytes(k)
+
+def find_rule(tables_format_rules, target_table_name):
+    for table_name_re, format_rules in tables_format_rules.items():
+        if re.compile(table_name_re).match(target_table_name):
```
Comment (eddyashton) [inline python/ccf/read_ledger.py:55]: Not critical, since we don't have any performance bounds for this, but recompiling the regex here every time will get very slow for large ledgers/many rules. Should be possible to precompile (convert a `list-of-pairs-of-(string-and-rule)` to a `list-of-pairs-of-(regex-and-rule)`, and pass that here.

File: python/ccf/read_ledger.py (hunk 3 above)
Comment (jumaffre) [reply python/ccf/read_ledger.py:50]: Yes, you're right, I'll change that (`dict`s are ordered since Python3.7 btw).

File: tests/suite/test_suite.py (hunk 2 above)
Comment (jumaffre) [reply tests/suite/test_suite.py]: I've added a comment to reflect this.

File: python/ccf/read_ledger.py [hunk 5]
```diff
@@ -14,18 +14,57 @@ def indent(n):
...
 
-def stringify_bytes(bs):
-    s = bs.decode()
-    if s.isprintable():
-        return s
-    if len(bs) > 0 and len(bs) <= 8:
-        n = int.from_bytes(bs, byteorder="little")
-        return f"<u{8 * len(bs)}: {n}>"
-    return bs
+def fmt_unsigned_int_little(data):
```
Comment (achamayou) [inline python/ccf/read_ledger.py:17]:
```suggestion
def fmt_uint64_le(data):
```

File: tests/infra/consortium.py [hunk 6]
```diff
@@ -364,7 +364,7 @@ def create_and_withdraw_large_proposal(self, remote_node):
         when trying to use ccf.ledger to read ledger entries.
         """
         proposal, _ = self.make_proposal(
-            "set_user", self.user_cert_path("user0"), {"padding": " " * 4096 * 5}
+            "set_user", self.user_cert_path("user0"), {"padding": "x" * 4096 * 5}
```
Comment (achamayou) [inline tests/infra/consortium.py:367]: what's the reason for this change?

**Review by achamayou — APPROVED** [review]

**Comment by achamayou** [comment]
@jumaffre this is a user-visible change, can we have a CHANGELOG entry please?
